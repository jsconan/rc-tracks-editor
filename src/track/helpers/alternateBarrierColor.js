/**
 * RC Tracks Editor
 * Copyright (c) 2022 Jean-SÃ©bastien CONAN
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */

import { alternate } from '../../core/helpers';

/**
 * A list of colors for the barriers.
 * @type {string[]}
 */
const barrierColors = ['even', 'odd'];

/**
 * Returns the color for a barrier element at the given position.
 * @param {number} i - The position of the barrier element.
 * @returns {string} - The color for the barrier element.
 */
export default i => alternate(i, barrierColors);
