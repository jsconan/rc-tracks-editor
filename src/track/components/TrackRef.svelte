<script>
    // Licensed under GNU Public License version 3
    // Copyright (c) 2022 Jean-SÃ©bastien CONAN

    import { createEventDispatcher } from 'svelte';
    import { Sketch } from '../elements';
    import { TILE_DIRECTION_RIGHT } from '../../tile/helpers';
    import { TileCoordList } from '../../tile/models';
    import { Tile } from '../../tile/components';

    export let list;
    export let x = 0;
    export let y = 0;
    export let width = void 0;
    export let height = void 0;

    TileCoordList.validateInstance(list);

    const dispatch = createEventDispatcher();

    /**
     * Handles the click event.
     * @param event
     * @private
     */
    function click(event) {
        const index = list.getIndex(event.target.dataset.id);
        if (index > -1) {
            const { id, type, direction, ratio, x, y, angle } = track.tiles[index];
            dispatch('click', { id, type, direction, ratio, x, y, angle });
        }
    }

    /**
     * Registers the tile models from the list.
     * @param {import('../../tile/models/TileCoordList.js').listCoord} list
     * @private
     */
    function registerModels(list) {
        const models = [];
        const registry = new Set();

        list.tiles.forEach(tile => {
            const { modelId } = tile.model;
            if (registry.has(modelId)) {
                return;
            }
            registry.add(modelId);

            const { model } = tile;
            models.push(model.clone().setDirection(TILE_DIRECTION_RIGHT));
        });

        return models;
    }

    $: track = $list;
    $: models = registerModels(track);
</script>

<Sketch
    {x}
    {y}
    {width}
    {height}
    viewX={track.x}
    viewY={track.y}
    viewWidth={track.width}
    viewHeight={track.height}
    on:click={click}
>
    {#each track.tiles as { id, x, y, angle, model } (id)}
        <use data-id={id} {x} {y} href="#{model.modelId}" transform={model.getRotateTransform(x, y, angle)} />
    {/each}
    <svelte:fragment slot="defs">
        {#each models as { id, type, ratio } (id)}
            <Tile {type} {ratio} {id} />
        {/each}
        <slot name="defs" />
    </svelte:fragment>
    <slot />
</Sketch>
