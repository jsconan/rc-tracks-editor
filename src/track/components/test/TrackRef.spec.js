/**
 * RC Tracks Editor
 * Copyright (c) 2022 Jean-SÃ©bastien CONAN
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */

import { render, fireEvent } from '@testing-library/svelte';
import Context from './Context.svelte';
import TrackRef from '../TrackRef.svelte';
import TrackWithSlot from './TrackWithSlot.svelte';
import { buildTrack } from '../../builders';
import { TileCoordList, TileList } from '../../models';
import { TileSpecifications } from '../../config';
import { CURVED_TILE_ENLARGED_TYPE, CURVED_TILE_TYPE, STRAIGHT_TILE_TYPE, TILE_DIRECTION_RIGHT } from '../../helpers';
import { tick } from 'svelte';

const laneWidth = 80;
const barrierWidth = 5;
const barrierChunks = 4;
const specs = new TileSpecifications(laneWidth, barrierWidth, barrierChunks);
const tileList = new TileList(specs);
tileList.import([
    { type: STRAIGHT_TILE_TYPE, direction: TILE_DIRECTION_RIGHT, ratio: 1 },
    { type: CURVED_TILE_TYPE, direction: TILE_DIRECTION_RIGHT, ratio: 1 },
    { type: CURVED_TILE_ENLARGED_TYPE, direction: TILE_DIRECTION_RIGHT, ratio: 1 }
]);
const listCoord = new TileCoordList(tileList, buildTrack, { hPadding: 10, vPadding: 20, startAngle: 90 });

describe('TrackRef', () => {
    it('renders with default values', () => {
        const props = { list: listCoord };
        const { container } = render(Context, {
            props: {
                component: TrackRef,
                contextKey: TileSpecifications.CONTEXT_ID,
                context: specs,
                props
            }
        });

        expect(container).toMatchSnapshot();
    });

    it('renders with the given parameters', () => {
        const props = {
            list: listCoord,
            x: 100,
            y: 200,
            width: 400,
            height: 400
        };
        const { container } = render(Context, {
            props: {
                component: TrackRef,
                contextKey: TileSpecifications.CONTEXT_ID,
                context: specs,
                props
            }
        });

        expect(container).toMatchSnapshot();
    });

    it('updates when the model is modified', async () => {
        const props = {
            list: listCoord,
            x: 100,
            y: 200,
            width: 400,
            height: 400
        };
        const rendered = render(Context, {
            props: {
                component: TrackRef,
                contextKey: TileSpecifications.CONTEXT_ID,
                context: specs,
                props
            }
        });

        await tick();
        tileList.append();
        await tick();
        expect(rendered.container).toMatchSnapshot();
    });

    it('needs a valid model', () => {
        expect(() =>
            render(TrackRef, {
                props: {
                    list: {}
                }
            })
        ).toThrow("'list' is not a store with a 'subscribe' method");

        expect(() =>
            render(TrackRef, {
                props: {
                    list: {
                        subscribe() {
                            return () => {};
                        }
                    }
                }
            })
        ).toThrow('The object must be an instance of TileCoordList!');
    });

    it('fires click', () => {
        const onClick = jest.fn();
        const props = { list: listCoord };
        const { container, component } = render(Context, {
            props: {
                component: TrackRef,
                contextKey: TileSpecifications.CONTEXT_ID,
                context: specs,
                props
            }
        });

        component.$on('click', onClick);
        fireEvent.click(container.querySelector(`[data-id=id-0]`));
        fireEvent.click(container.querySelector(`[data-id=id-1]`));
        fireEvent.click(container.querySelector(`[data-id=id-2]`));
        expect(onClick).toHaveBeenCalledTimes(3);
    });

    it('renders with the given element in slots', () => {
        const props = { list: listCoord, component: TrackRef };
        const { container } = render(Context, {
            props: {
                component: TrackWithSlot,
                contextKey: TileSpecifications.CONTEXT_ID,
                context: specs,
                props
            }
        });

        expect(container).toMatchSnapshot();
    });
});
