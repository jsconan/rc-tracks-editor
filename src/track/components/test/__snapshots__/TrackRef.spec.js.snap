// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TrackRef renders with default values 1`] = `
<body>
  <div>
    <svg
      viewBox="-220 -175 230 240"
      x="0"
      y="0"
    >
      <defs>
        
        <g
          class="tile straight-tile"
          id="straight-tile-1"
          transform=""
        >
          <rect
            class="ground"
            height="110"
            width="90"
            x="-45"
            y="0"
          />
          <g
            class="barrier straight-barrier"
          >
            
            <rect
              class="barrier-chunk even"
              height="27.5"
              width="5"
              x="-45"
              y="0"
            />
            
            <rect
              class="barrier-chunk odd"
              height="27.5"
              width="5"
              x="-45"
              y="27.5"
            />
            
            <rect
              class="barrier-chunk even"
              height="27.5"
              width="5"
              x="-45"
              y="55"
            />
            
            <rect
              class="barrier-chunk odd"
              height="27.5"
              width="5"
              x="-45"
              y="82.5"
            />
          </g>
          <g
            class="barrier straight-barrier"
          >
            
            <rect
              class="barrier-chunk odd"
              height="27.5"
              width="5"
              x="40"
              y="0"
            />
            
            <rect
              class="barrier-chunk even"
              height="27.5"
              width="5"
              x="40"
              y="27.5"
            />
            
            <rect
              class="barrier-chunk odd"
              height="27.5"
              width="5"
              x="40"
              y="55"
            />
            
            <rect
              class="barrier-chunk even"
              height="27.5"
              width="5"
              x="40"
              y="82.5"
            />
          </g>
        </g>
        
        
        <g
          class="tile curved-tile"
          id="curved-tile-1"
          transform=""
        >
          <path
            class="ground"
            d="M -45 0 A 10 10 0 0 1 -55 10 L -54.99999999999999 100 A 100 100 0 0 0 45 0 Z"
          />
          <g
            class="barrier curved-barrier"
          >
            
            <path
              class="barrier-chunk even"
              d="M -45 0 A 10 10 0 0 1 -47.928932188134524 7.071067811865475 L -44.39339828220179 10.606601717798211 A 15 15 0 0 0 -40 0 Z"
            />
            
            <path
              class="barrier-chunk odd"
              d="M -47.928932188134524 7.071067811865475 A 10 10 0 0 1 -55 10 L -55 15 A 15 15 0 0 0 -44.39339828220179 10.606601717798211 Z"
            />
          </g>
          <g
            class="barrier curved-barrier"
          >
            
            <path
              class="barrier-chunk odd"
              d="M 40 0 A 95 95 0 0 1 32.76855558857224 36.354926074683526 L 37.38795325112868 38.268343236508976 A 100 100 0 0 0 45 0 Z"
            />
            
            <path
              class="barrier-chunk even"
              d="M 32.76855558857224 36.354926074683526 A 95 95 0 0 1 12.175144212722017 67.175144212722 L 15.710678118654755 70.71067811865474 A 100 100 0 0 0 37.38795325112868 38.268343236508976 Z"
            />
            
            <path
              class="barrier-chunk odd"
              d="M 12.175144212722017 67.175144212722 A 95 95 0 0 1 -18.645073925316467 87.76855558857224 L -16.731656763491017 92.38795325112868 A 100 100 0 0 0 15.710678118654755 70.71067811865474 Z"
            />
            
            <path
              class="barrier-chunk even"
              d="M -18.645073925316467 87.76855558857224 A 95 95 0 0 1 -54.99999999999999 95 L -54.99999999999999 100 A 100 100 0 0 0 -16.731656763491017 92.38795325112868 Z"
            />
          </g>
        </g>
        
        
        <g
          class="tile curved-tile-enlarged"
          id="curved-tile-enlarged-1"
          transform=""
        >
          <path
            class="ground"
            d="M -45 0 A 10 10 0 0 1 -55 10 L -55 100 L 0 100 A 45 45 0 0 0 45 55 L 45 0 Z"
          />
          <g
            class="barrier curved-barrier"
          >
            
            <path
              class="barrier-chunk even"
              d="M -45 0 A 10 10 0 0 1 -47.928932188134524 7.071067811865475 L -44.39339828220179 10.606601717798211 A 15 15 0 0 0 -40 0 Z"
            />
            
            <path
              class="barrier-chunk odd"
              d="M -47.928932188134524 7.071067811865475 A 10 10 0 0 1 -55 10 L -55 15 A 15 15 0 0 0 -44.39339828220179 10.606601717798211 Z"
            />
          </g>
          <g
            class="barrier curved-barrier"
          >
            
            <path
              class="barrier-chunk odd"
              d="M 40 55 A 40 40 0 0 1 28.284271247461902 83.2842712474619 L 31.81980515339464 86.81980515339464 A 45 45 0 0 0 45 55 Z"
            />
            
            <path
              class="barrier-chunk even"
              d="M 28.284271247461902 83.2842712474619 A 40 40 0 0 1 2.4492935982947065e-15 95 L 2.7554552980815448e-15 100 A 45 45 0 0 0 31.81980515339464 86.81980515339464 Z"
            />
          </g>
          <g
            class="barrier straight-barrier"
          >
            
            <rect
              class="barrier-chunk even"
              height="5"
              width="27.5"
              x="-55"
              y="95"
            />
            
            <rect
              class="barrier-chunk odd"
              height="5"
              width="27.5"
              x="-27.5"
              y="95"
            />
          </g>
          <g
            class="barrier straight-barrier"
          >
            
            <rect
              class="barrier-chunk odd"
              height="27.5"
              width="5"
              x="40"
              y="0"
            />
            
            <rect
              class="barrier-chunk even"
              height="27.5"
              width="5"
              x="40"
              y="27.5"
            />
          </g>
        </g>
        
         
      </defs>
      
      <use
        data-id="id-0"
        href="#straight-tile-1"
        transform="rotate(90 0 0)"
        x="0"
        y="0"
      />
      <use
        data-id="id-1"
        href="#curved-tile-1"
        transform="rotate(90 -110 6.735557395310443e-15)"
        x="-110"
        y="6.735557395310443e-15"
      />
      <use
        data-id="id-2"
        href="#curved-tile-enlarged-1"
        transform="rotate(180 -165 -54.99999999999999)"
        x="-165"
        y="-54.99999999999999"
      />
       
    </svg>
    
  </div>
</body>
`;

exports[`TrackRef renders with the given element in slots 1`] = `
<body>
  <div />
  <div />
  <div>
    <svg
      viewBox="-220 -175.00000000000003 230 240.00000000000003"
      x="0"
      y="0"
    >
      <defs>
        
        <g
          class="tile straight-tile"
          id="straight-tile-1"
          transform=""
        >
          <rect
            class="ground"
            height="110"
            width="90"
            x="-45"
            y="0"
          />
          <g
            class="barrier straight-barrier"
          >
            
            <rect
              class="barrier-chunk even"
              height="27.5"
              width="5"
              x="-45"
              y="0"
            />
            
            <rect
              class="barrier-chunk odd"
              height="27.5"
              width="5"
              x="-45"
              y="27.5"
            />
            
            <rect
              class="barrier-chunk even"
              height="27.5"
              width="5"
              x="-45"
              y="55"
            />
            
            <rect
              class="barrier-chunk odd"
              height="27.5"
              width="5"
              x="-45"
              y="82.5"
            />
          </g>
          <g
            class="barrier straight-barrier"
          >
            
            <rect
              class="barrier-chunk odd"
              height="27.5"
              width="5"
              x="40"
              y="0"
            />
            
            <rect
              class="barrier-chunk even"
              height="27.5"
              width="5"
              x="40"
              y="27.5"
            />
            
            <rect
              class="barrier-chunk odd"
              height="27.5"
              width="5"
              x="40"
              y="55"
            />
            
            <rect
              class="barrier-chunk even"
              height="27.5"
              width="5"
              x="40"
              y="82.5"
            />
          </g>
        </g>
        
        
        <g
          class="tile curved-tile"
          id="curved-tile-1"
          transform=""
        >
          <path
            class="ground"
            d="M -45 0 A 10 10 0 0 1 -55 10 L -54.99999999999999 100 A 100 100 0 0 0 45 0 Z"
          />
          <g
            class="barrier curved-barrier"
          >
            
            <path
              class="barrier-chunk even"
              d="M -45 0 A 10 10 0 0 1 -47.928932188134524 7.071067811865475 L -44.39339828220179 10.606601717798211 A 15 15 0 0 0 -40 0 Z"
            />
            
            <path
              class="barrier-chunk odd"
              d="M -47.928932188134524 7.071067811865475 A 10 10 0 0 1 -55 10 L -55 15 A 15 15 0 0 0 -44.39339828220179 10.606601717798211 Z"
            />
          </g>
          <g
            class="barrier curved-barrier"
          >
            
            <path
              class="barrier-chunk odd"
              d="M 40 0 A 95 95 0 0 1 32.76855558857224 36.354926074683526 L 37.38795325112868 38.268343236508976 A 100 100 0 0 0 45 0 Z"
            />
            
            <path
              class="barrier-chunk even"
              d="M 32.76855558857224 36.354926074683526 A 95 95 0 0 1 12.175144212722017 67.175144212722 L 15.710678118654755 70.71067811865474 A 100 100 0 0 0 37.38795325112868 38.268343236508976 Z"
            />
            
            <path
              class="barrier-chunk odd"
              d="M 12.175144212722017 67.175144212722 A 95 95 0 0 1 -18.645073925316467 87.76855558857224 L -16.731656763491017 92.38795325112868 A 100 100 0 0 0 15.710678118654755 70.71067811865474 Z"
            />
            
            <path
              class="barrier-chunk even"
              d="M -18.645073925316467 87.76855558857224 A 95 95 0 0 1 -54.99999999999999 95 L -54.99999999999999 100 A 100 100 0 0 0 -16.731656763491017 92.38795325112868 Z"
            />
          </g>
        </g>
        
        
        <g
          class="tile curved-tile-enlarged"
          id="curved-tile-enlarged-1"
          transform=""
        >
          <path
            class="ground"
            d="M -45 0 A 10 10 0 0 1 -55 10 L -55 100 L 0 100 A 45 45 0 0 0 45 55 L 45 0 Z"
          />
          <g
            class="barrier curved-barrier"
          >
            
            <path
              class="barrier-chunk even"
              d="M -45 0 A 10 10 0 0 1 -47.928932188134524 7.071067811865475 L -44.39339828220179 10.606601717798211 A 15 15 0 0 0 -40 0 Z"
            />
            
            <path
              class="barrier-chunk odd"
              d="M -47.928932188134524 7.071067811865475 A 10 10 0 0 1 -55 10 L -55 15 A 15 15 0 0 0 -44.39339828220179 10.606601717798211 Z"
            />
          </g>
          <g
            class="barrier curved-barrier"
          >
            
            <path
              class="barrier-chunk odd"
              d="M 40 55 A 40 40 0 0 1 28.284271247461902 83.2842712474619 L 31.81980515339464 86.81980515339464 A 45 45 0 0 0 45 55 Z"
            />
            
            <path
              class="barrier-chunk even"
              d="M 28.284271247461902 83.2842712474619 A 40 40 0 0 1 2.4492935982947065e-15 95 L 2.7554552980815448e-15 100 A 45 45 0 0 0 31.81980515339464 86.81980515339464 Z"
            />
          </g>
          <g
            class="barrier straight-barrier"
          >
            
            <rect
              class="barrier-chunk even"
              height="5"
              width="27.5"
              x="-55"
              y="95"
            />
            
            <rect
              class="barrier-chunk odd"
              height="5"
              width="27.5"
              x="-27.5"
              y="95"
            />
          </g>
          <g
            class="barrier straight-barrier"
          >
            
            <rect
              class="barrier-chunk odd"
              height="27.5"
              width="5"
              x="40"
              y="0"
            />
            
            <rect
              class="barrier-chunk even"
              height="27.5"
              width="5"
              x="40"
              y="27.5"
            />
          </g>
        </g>
        
         
        <circle
          cx="10"
          cy="10"
          id="circle"
          r="10"
          slot="defs"
        />
      </defs>
      
      <use
        data-id="id-0"
        href="#straight-tile-1"
        transform="rotate(90 0 0)"
        x="0"
        y="0"
      />
      <use
        data-id="id-1"
        href="#curved-tile-1"
        transform="rotate(90 -110 6.735557395310443e-15)"
        x="-110"
        y="6.735557395310443e-15"
      />
      <use
        data-id="id-2"
        href="#curved-tile-enlarged-1"
        transform="rotate(180 -165 -54.99999999999999)"
        x="-165"
        y="-54.99999999999999"
      />
      <use
        data-id="id-3"
        href="#straight-tile-1"
        transform="rotate(270 -110 -110)"
        x="-110"
        y="-110"
      />
       
      <rect
        height="50"
        width="100"
        x="10"
        y="10"
      />
    </svg>
    
    
  </div>
</body>
`;

exports[`TrackRef renders with the given parameters 1`] = `
<body>
  <div>
    <svg
      height="400"
      viewBox="-220 -175 230 240"
      width="400"
      x="100"
      y="200"
    >
      <defs>
        
        <g
          class="tile straight-tile"
          id="straight-tile-1"
          transform=""
        >
          <rect
            class="ground"
            height="110"
            width="90"
            x="-45"
            y="0"
          />
          <g
            class="barrier straight-barrier"
          >
            
            <rect
              class="barrier-chunk even"
              height="27.5"
              width="5"
              x="-45"
              y="0"
            />
            
            <rect
              class="barrier-chunk odd"
              height="27.5"
              width="5"
              x="-45"
              y="27.5"
            />
            
            <rect
              class="barrier-chunk even"
              height="27.5"
              width="5"
              x="-45"
              y="55"
            />
            
            <rect
              class="barrier-chunk odd"
              height="27.5"
              width="5"
              x="-45"
              y="82.5"
            />
          </g>
          <g
            class="barrier straight-barrier"
          >
            
            <rect
              class="barrier-chunk odd"
              height="27.5"
              width="5"
              x="40"
              y="0"
            />
            
            <rect
              class="barrier-chunk even"
              height="27.5"
              width="5"
              x="40"
              y="27.5"
            />
            
            <rect
              class="barrier-chunk odd"
              height="27.5"
              width="5"
              x="40"
              y="55"
            />
            
            <rect
              class="barrier-chunk even"
              height="27.5"
              width="5"
              x="40"
              y="82.5"
            />
          </g>
        </g>
        
        
        <g
          class="tile curved-tile"
          id="curved-tile-1"
          transform=""
        >
          <path
            class="ground"
            d="M -45 0 A 10 10 0 0 1 -55 10 L -54.99999999999999 100 A 100 100 0 0 0 45 0 Z"
          />
          <g
            class="barrier curved-barrier"
          >
            
            <path
              class="barrier-chunk even"
              d="M -45 0 A 10 10 0 0 1 -47.928932188134524 7.071067811865475 L -44.39339828220179 10.606601717798211 A 15 15 0 0 0 -40 0 Z"
            />
            
            <path
              class="barrier-chunk odd"
              d="M -47.928932188134524 7.071067811865475 A 10 10 0 0 1 -55 10 L -55 15 A 15 15 0 0 0 -44.39339828220179 10.606601717798211 Z"
            />
          </g>
          <g
            class="barrier curved-barrier"
          >
            
            <path
              class="barrier-chunk odd"
              d="M 40 0 A 95 95 0 0 1 32.76855558857224 36.354926074683526 L 37.38795325112868 38.268343236508976 A 100 100 0 0 0 45 0 Z"
            />
            
            <path
              class="barrier-chunk even"
              d="M 32.76855558857224 36.354926074683526 A 95 95 0 0 1 12.175144212722017 67.175144212722 L 15.710678118654755 70.71067811865474 A 100 100 0 0 0 37.38795325112868 38.268343236508976 Z"
            />
            
            <path
              class="barrier-chunk odd"
              d="M 12.175144212722017 67.175144212722 A 95 95 0 0 1 -18.645073925316467 87.76855558857224 L -16.731656763491017 92.38795325112868 A 100 100 0 0 0 15.710678118654755 70.71067811865474 Z"
            />
            
            <path
              class="barrier-chunk even"
              d="M -18.645073925316467 87.76855558857224 A 95 95 0 0 1 -54.99999999999999 95 L -54.99999999999999 100 A 100 100 0 0 0 -16.731656763491017 92.38795325112868 Z"
            />
          </g>
        </g>
        
        
        <g
          class="tile curved-tile-enlarged"
          id="curved-tile-enlarged-1"
          transform=""
        >
          <path
            class="ground"
            d="M -45 0 A 10 10 0 0 1 -55 10 L -55 100 L 0 100 A 45 45 0 0 0 45 55 L 45 0 Z"
          />
          <g
            class="barrier curved-barrier"
          >
            
            <path
              class="barrier-chunk even"
              d="M -45 0 A 10 10 0 0 1 -47.928932188134524 7.071067811865475 L -44.39339828220179 10.606601717798211 A 15 15 0 0 0 -40 0 Z"
            />
            
            <path
              class="barrier-chunk odd"
              d="M -47.928932188134524 7.071067811865475 A 10 10 0 0 1 -55 10 L -55 15 A 15 15 0 0 0 -44.39339828220179 10.606601717798211 Z"
            />
          </g>
          <g
            class="barrier curved-barrier"
          >
            
            <path
              class="barrier-chunk odd"
              d="M 40 55 A 40 40 0 0 1 28.284271247461902 83.2842712474619 L 31.81980515339464 86.81980515339464 A 45 45 0 0 0 45 55 Z"
            />
            
            <path
              class="barrier-chunk even"
              d="M 28.284271247461902 83.2842712474619 A 40 40 0 0 1 2.4492935982947065e-15 95 L 2.7554552980815448e-15 100 A 45 45 0 0 0 31.81980515339464 86.81980515339464 Z"
            />
          </g>
          <g
            class="barrier straight-barrier"
          >
            
            <rect
              class="barrier-chunk even"
              height="5"
              width="27.5"
              x="-55"
              y="95"
            />
            
            <rect
              class="barrier-chunk odd"
              height="5"
              width="27.5"
              x="-27.5"
              y="95"
            />
          </g>
          <g
            class="barrier straight-barrier"
          >
            
            <rect
              class="barrier-chunk odd"
              height="27.5"
              width="5"
              x="40"
              y="0"
            />
            
            <rect
              class="barrier-chunk even"
              height="27.5"
              width="5"
              x="40"
              y="27.5"
            />
          </g>
        </g>
        
         
      </defs>
      
      <use
        data-id="id-0"
        href="#straight-tile-1"
        transform="rotate(90 0 0)"
        x="0"
        y="0"
      />
      <use
        data-id="id-1"
        href="#curved-tile-1"
        transform="rotate(90 -110 6.735557395310443e-15)"
        x="-110"
        y="6.735557395310443e-15"
      />
      <use
        data-id="id-2"
        href="#curved-tile-enlarged-1"
        transform="rotate(180 -165 -54.99999999999999)"
        x="-165"
        y="-54.99999999999999"
      />
       
    </svg>
    
  </div>
</body>
`;

exports[`TrackRef updates when the model is modified 1`] = `
<body>
  <div>
    <svg
      height="400"
      viewBox="-220 -175.00000000000003 230 240.00000000000003"
      width="400"
      x="100"
      y="200"
    >
      <defs>
        
        <g
          class="tile straight-tile"
          id="straight-tile-1"
          transform=""
        >
          <rect
            class="ground"
            height="110"
            width="90"
            x="-45"
            y="0"
          />
          <g
            class="barrier straight-barrier"
          >
            
            <rect
              class="barrier-chunk even"
              height="27.5"
              width="5"
              x="-45"
              y="0"
            />
            
            <rect
              class="barrier-chunk odd"
              height="27.5"
              width="5"
              x="-45"
              y="27.5"
            />
            
            <rect
              class="barrier-chunk even"
              height="27.5"
              width="5"
              x="-45"
              y="55"
            />
            
            <rect
              class="barrier-chunk odd"
              height="27.5"
              width="5"
              x="-45"
              y="82.5"
            />
          </g>
          <g
            class="barrier straight-barrier"
          >
            
            <rect
              class="barrier-chunk odd"
              height="27.5"
              width="5"
              x="40"
              y="0"
            />
            
            <rect
              class="barrier-chunk even"
              height="27.5"
              width="5"
              x="40"
              y="27.5"
            />
            
            <rect
              class="barrier-chunk odd"
              height="27.5"
              width="5"
              x="40"
              y="55"
            />
            
            <rect
              class="barrier-chunk even"
              height="27.5"
              width="5"
              x="40"
              y="82.5"
            />
          </g>
        </g>
        
        
        <g
          class="tile curved-tile"
          id="curved-tile-1"
          transform=""
        >
          <path
            class="ground"
            d="M -45 0 A 10 10 0 0 1 -55 10 L -54.99999999999999 100 A 100 100 0 0 0 45 0 Z"
          />
          <g
            class="barrier curved-barrier"
          >
            
            <path
              class="barrier-chunk even"
              d="M -45 0 A 10 10 0 0 1 -47.928932188134524 7.071067811865475 L -44.39339828220179 10.606601717798211 A 15 15 0 0 0 -40 0 Z"
            />
            
            <path
              class="barrier-chunk odd"
              d="M -47.928932188134524 7.071067811865475 A 10 10 0 0 1 -55 10 L -55 15 A 15 15 0 0 0 -44.39339828220179 10.606601717798211 Z"
            />
          </g>
          <g
            class="barrier curved-barrier"
          >
            
            <path
              class="barrier-chunk odd"
              d="M 40 0 A 95 95 0 0 1 32.76855558857224 36.354926074683526 L 37.38795325112868 38.268343236508976 A 100 100 0 0 0 45 0 Z"
            />
            
            <path
              class="barrier-chunk even"
              d="M 32.76855558857224 36.354926074683526 A 95 95 0 0 1 12.175144212722017 67.175144212722 L 15.710678118654755 70.71067811865474 A 100 100 0 0 0 37.38795325112868 38.268343236508976 Z"
            />
            
            <path
              class="barrier-chunk odd"
              d="M 12.175144212722017 67.175144212722 A 95 95 0 0 1 -18.645073925316467 87.76855558857224 L -16.731656763491017 92.38795325112868 A 100 100 0 0 0 15.710678118654755 70.71067811865474 Z"
            />
            
            <path
              class="barrier-chunk even"
              d="M -18.645073925316467 87.76855558857224 A 95 95 0 0 1 -54.99999999999999 95 L -54.99999999999999 100 A 100 100 0 0 0 -16.731656763491017 92.38795325112868 Z"
            />
          </g>
        </g>
        
        
        <g
          class="tile curved-tile-enlarged"
          id="curved-tile-enlarged-1"
          transform=""
        >
          <path
            class="ground"
            d="M -45 0 A 10 10 0 0 1 -55 10 L -55 100 L 0 100 A 45 45 0 0 0 45 55 L 45 0 Z"
          />
          <g
            class="barrier curved-barrier"
          >
            
            <path
              class="barrier-chunk even"
              d="M -45 0 A 10 10 0 0 1 -47.928932188134524 7.071067811865475 L -44.39339828220179 10.606601717798211 A 15 15 0 0 0 -40 0 Z"
            />
            
            <path
              class="barrier-chunk odd"
              d="M -47.928932188134524 7.071067811865475 A 10 10 0 0 1 -55 10 L -55 15 A 15 15 0 0 0 -44.39339828220179 10.606601717798211 Z"
            />
          </g>
          <g
            class="barrier curved-barrier"
          >
            
            <path
              class="barrier-chunk odd"
              d="M 40 55 A 40 40 0 0 1 28.284271247461902 83.2842712474619 L 31.81980515339464 86.81980515339464 A 45 45 0 0 0 45 55 Z"
            />
            
            <path
              class="barrier-chunk even"
              d="M 28.284271247461902 83.2842712474619 A 40 40 0 0 1 2.4492935982947065e-15 95 L 2.7554552980815448e-15 100 A 45 45 0 0 0 31.81980515339464 86.81980515339464 Z"
            />
          </g>
          <g
            class="barrier straight-barrier"
          >
            
            <rect
              class="barrier-chunk even"
              height="5"
              width="27.5"
              x="-55"
              y="95"
            />
            
            <rect
              class="barrier-chunk odd"
              height="5"
              width="27.5"
              x="-27.5"
              y="95"
            />
          </g>
          <g
            class="barrier straight-barrier"
          >
            
            <rect
              class="barrier-chunk odd"
              height="27.5"
              width="5"
              x="40"
              y="0"
            />
            
            <rect
              class="barrier-chunk even"
              height="27.5"
              width="5"
              x="40"
              y="27.5"
            />
          </g>
        </g>
        
         
      </defs>
      
      <use
        data-id="id-0"
        href="#straight-tile-1"
        transform="rotate(90 0 0)"
        x="0"
        y="0"
      />
      <use
        data-id="id-1"
        href="#curved-tile-1"
        transform="rotate(90 -110 6.735557395310443e-15)"
        x="-110"
        y="6.735557395310443e-15"
      />
      <use
        data-id="id-2"
        href="#curved-tile-enlarged-1"
        transform="rotate(180 -165 -54.99999999999999)"
        x="-165"
        y="-54.99999999999999"
      />
      <use
        data-id="id-3"
        href="#straight-tile-1"
        transform="rotate(270 -110 -110)"
        x="-110"
        y="-110"
      />
       
    </svg>
    
  </div>
</body>
`;
